pcilmr.man
.TH PCILMR 8 "@TODAY@" "@VERSION@" "The PCI Utilities"
.SH NAME
pcilmr \- margin PCIe Links
.SH SYNOPSIS
.B pcilmr
.RB [ "--margin" ]
.RI [ "<common options>" ] " <link port> " [ "<link options>" "] [" "<link port> " [ "<link options>" ] " ..." ]
.br
.B pcilmr --full
.RI [ "<common options>" ]
.br
.B pcilmr --scan
.SH CONFIGURATION
List of the requirements for links and system settings
to run the margining test.

.B BIOS settings
(depends on the system, relevant for server baseboards
with Xeon CPUs):
.IP \[bu] 3
Turn off PCIe Leaky Bucket Feature, Re-Equalization and Link Degradation;
.IP \[bu]
Set Error Thresholds to 0;
.IP \[bu]
Intel VMD for NVMe SSDs - in case of strange behavior of the
.BR pcilmr,
try to run it with the VMD turned off.
.PP
.B Device (link) requirements:
.IP
.I "Configured by the user before running the utility, the utility does not change them:"
.RS
.IP \[bu] 3
The current Link data rate must be 16.0 GT/s or higher (right now
utility supports 16 GT/s and 32 GT/s Links);
.IP \[bu]
Link Downstream Component must be at D0 Power Management State.
.RE
.IP
.I "Configured by the utility during operation, utility set them to their original "
.I "state after receiving the results:"
.RS
.IP \[bu] 3
The ASPM must be disabled in both the Downstream Port and Upstream Port;
.IP \[bu]
The Hardware Autonomous Speed Disable bit of the Link Control 2 register must be Set in both the
Downstream Port and Upstream Port;
.IP \[bu]
The Hardware Autonomous Width Disable bit of the Link Control register must be Set in both the
Downstream Port and Upstream Port.
.SH DESCRIPTION
.B pcilmr
utility allows you to take advantage of the PCIe Lane Margining at the Receiver
capability which is mandatory for all Ports supporting a data rate of 16.0 GT/s or
higher, including Pseudo Ports (Retimers). Lane Margining at Receiver enables system
software to obtain the margin information of a given Receiver while the Link is in the
L0 state. The margin information includes both voltage and time, in either direction from
the current Receiver position. Margining support for timing is required, while support
for voltage is optional at 16.0 GT/s and required at 32.0 GT/s and higher data rates. Also,
independent time margining and independent voltage margining is optional.

Utility allows to get an approximation of the eye margin diagram in the form of a rhombus
(by four points). Lane Margining at the Receiver capability enables users to margin PCIe
links without a hardware debugger and without the need to stop the target system. Utility
can be useful to debug link issues due to receiver margins.

.B pcilmr
requires root privileges (to access Extended Configuration Space), but during our testing
there were no problems with the devices and they successfully returned to their normal initial
state after the end of testing.

.SH RESULTS GRADING
The PCIe specification provides reference values for the eye diagram, which are also used by the
.B pcilmr
to evaluate the results, but it seems that it makes sense to contact the
manufacturer of a particular device for references.

The utility uses values set in PCIe Base Spec Rev. 5.0 Section 8.4.2 as the default eye width and height
minimum references. Recommended values were taken from
the PCIe Architecture PHY Test Spec Rev 5.0 (Transmitter Electrical Compliance).

Reference grading values currently used by the utility are presented in the table below:

.TS
box tab(:);
C | Cb S | Cb S
C | Cb | Cb | Cb | Cb
Lb | C | C | C | C.
\&:16 GT/s (Gen 4):32 GT/s (Gen 5)
\&:EW:EH:EW:EH
_
Min:T{
18.75 ps
.br
30% UI
T}:15 mV:T{
9.375 ps
.br
30% UI
T}:15 mV
_
Rec:T{
23.75 ps
.br
38% UI
T}:21 mV:T{
10.157 ps
.br
33% UI
T}:19.75 mV
.TE

.B pcilmr
uses full eye width and height values to grade lanes. However, it is possible that
device supports only one side margining. In such cases by default utility will
calculate EW or EH as a double one side result.

If info for specific device is available, you can configure grading criteria
and tweak utility behavior in one-side only cases for your device using
.I -g
link specific option (see below).

.SH HARDWARE QUIRKS SUPPORT

Thanks to testing or directly from the manufacturer's documentation, we know that
some devices require special treatment during the margining.
Utility detects such devices based on their Vendor ID - Device ID pair.
Right now the list of special devices is hardcoded in
.I margin_hw
file. For such devices utility can automatically adjust port margining parameters
or grading options.

For example, for Ice Lake CPUs RC ports
.B pcilmr
will change device MaxVoltageOffset value and will force the use of
.RI ' "one side is the whole" "' grading mode."

.SH OPTIONS
.SS Device Specifier
.B "You can specify Downstream or Upstream Port of the Link."
.TP
.B "<device/component>" \t
.RI [ "<domain>" :] <bus> : <dev> . <func>
(see
.BR lspci (8))
.SS Utility Modes
.TP
.BI --margin " <downstream component> ..."
Margin selected Links.
.TP
.B --full
Margin all ready for testing (in a meaning similar to the
.B --scan
option) Links in the system (one by one).
.TP
.B --scan
Scan for Links with negotiated speed 16 GT/s or higher. Mark "Ready" those of them
in which at least one of the Link sides have Margining Ready bit set meaning that
these Links are ready for testing and you can run utility on them.
.SS Margining Common (for all specified links) options
.B -c
Print Device Lane Margining Capabilities only. Do not run margining.
.TP
.BI -e " <errors>"
Specify Error Count Limit for margining.
.br
Default: 4.
.TP
.BI -o " <directory>"
Save margining results in csv form into the specified directory. Utility
will generate file with the name in form of
.RI "\[dq]lmr_" "<port>" "_Rx" # _ <timestamp> ".csv\[dq]"
for each successfully tested receiver.
.TP
.BI -d " <time>"
Specify dwell time in seconds for the margining step.
.br
Default: 1 s
.SS Margining Link specific options
.TP
\fB\-l\fI <lane>\fP[\fI,<lane>...\fP]
.R Specify lanes for margining.
.br
Remember that Device may use Lane Reversal for Lane numbering. However, utility
uses logical lane numbers in arguments and for logging. Utility will automatically
determine Lane Reversal and tune its calls.
.br
Default: all link lanes.
.TP
\fB-r\fI <recvn>\fP[\fI,<recvn>...\fP]
Specify Receivers to select margining targets.
.br
Default: all available Receivers (including Retimers).
.TP
.BI -p " <parallel_lanes>"
Specify number of lanes to margin simultaneously.
.br
According to spec it's possible for Receiver to margin up to MaxLanes + 1
lanes simultaneously, but during testing, performing margining on several
lanes simultaneously led to results that were different from sequential
margining, so this feature requires additional verification and
.I -p
option right now is for experiments mostly.
.br
Default: 1.
.PP
.B "Use only one of -T/-t options at the same time (same for -V/-v)."
.br
.B "Without these options utility will use MaxSteps from Device"
.B "capabilities as test limit."
.TP
.B -T
Time Margining will continue until the Error Count is no more
than an Error Count Limit. Use this option to find Link limit.
.TP
.BI -t " <steps>"
Specify maximum number of steps for Time Margining.
.TP
.B -V
Same as
.I -T
option, but for Voltage.
.TP
.BI -v " <steps>"
Specify maximum number of steps for Voltage Margining.
.TP
\fB-g\fI <recvn>\fPt=\fI<criteria>\fP{%|ps}[,f]
.TP
.IB "   <recvn>" t=f[, "<criteria>" "{%|ps}]"
.TP
.IB "   <recvn>" v= "<criteria>" "[,f]"
.TP
.IB "   <recvn>" v=f[, "<criteria>" ]
Specify pass/fail grading criteria for eye width (timing - t) or height
(voltage - v) for one of the link receivers. For EW you must choose one of the
units (% UI or ps), for EH mV is used.
.br
Additional flag
.I f
is for situations when port doesn't support two side independent
margining. In such cases by default utility will calculate EW or EH as a
double one side result. You can add
.I f
flag for
.I -g
option to tell the utility that the result in one direction is actually the
measurement of the full eye and it does not need to be multiplied. This is so called
.RI ' "one side is the whole" "' grading mode."

.SH EXAMPLES
Utility syntax example:
.RS
.BI "pcilmr -o " "csv ab:0.0 " "-r " "1,6 " "-g " "1t=20% " "-g " "1v=f,30 52:0.0 " "-l " "0,1,2 " "-TV"
.RE

.UR https://gist.github.com/bombanya/f2b15263712757ffba1a11eea011c419
Examples of collected results on different systems.
.UE

.SH SEE ALSO
.nh
.BR lspci (8),
.B PCI Express Base Specification (Lane Margining at Receiver)
.hy


ChangeLog
2024-05-30  Martin Mares <mj@ucw.cz>

	* Released as 3.13.0.

	* lspci decodes CXL 1.1 device link status information.
	  This requires a recent kernel which exports rcd_* atributes via
	  sysfs.

	* Further development of the pcilmr (the link margining utility)

	* Dump parsing supports 6-digit domain numbers.

	* Bug fixes in PCIe link state reporting.

	* Decode more fields in PCIe AER capability.

	* Fixed build on Linux systems with musl libc.

	* Updated pci.ids.

2024-04-05  Martin Mares <mj@ucw.cz>

	* Released as 3.12.0.

	* lspci decodes the IDE (Integrity & Data Encryption) and TEE-IO
	  extended capabilities.

	* Several bugs in back-ends for Windows were fixed and the
	  README.Windows updated.

	* Fixed building on Haiku.

	* We now use more C99 features.

	* Optimization flags used for compiling individual object files
	  should be the same as optimization flags for linking the final
	  executable to make link-time optimization possible.

	* We no longer look up subsystems in the HWDB, because there is
	  no API for querying only the subsystem. The lookup we used
	  previously returned a device name if there was no entry for
	  the particular subsystem.

	* Updated pci.ids.

2024-02-25  Martin Mares <mj@ucw.cz>

	* Released as 3.11.1.

	* Fixed wrong API version in lib/pci.h.

	* Updated README.Windows.

	* Fix compilation on Windows.

2024-02-24  Martin Mares <mj@ucw.cz>

	* Released as 3.11.0.

	* update-pciids now supports XZ compression. If libpci is configured
	  with support for compression, all downloaded files are recompressed
	  as gzip. Otherwise they are stored as plain text.

	* update-pciids now sends itself as the User-Agent.

	* Added a pcilmr utility for PCIe lane margining. Thanks to Nikita
	  Proshkin for contributing it.

	* Re-factored access to i386 ports on all relevant platforms.

	* Added i386 port access on OpenBSD.

	* Back-ends for Windows received many bug fixes and improvements.

	* ECAM back-end now scans ACPI and BIOS memory faster.

	* Linux systems without pread/pwrite are no longer supported
	  as they are hopefully long gone. This helps avoid the tricky check
	  for presence of pread which was found to fail on musl libc.

	* Improved decoding of PCIe control and status registers.

	* Decoding of CXL capabilities now supports up to CXL 3.0.

	* lspci now displays interrupt message numbers consistently across
	  different capabilities.

	* Cache of IDs resolved via DNS, which was located in ~/.pci-ids
	  by default, is now stored according to the XDG base directory
	  specification in $XDG_CACHE_HOME/pci-ids.

	* All source files now have SPDX license identifiers.

	* Internal: The "aux" fields of structs pci_access and pci_dev
	  reserved for use by back-ends were renamed to backend_data to better
	  reflect their meaning.

	* As usually, various minor bug fixes and updated pci.ids.

2023-05-01  Martin Mares <mj@ucw.cz>

	* Released as 3.10.0.

	* Fixed bug in definition of versioned symbol aliases
	  in shared libpci, which made compiling with link-time
	  optimization fail.

	* Filters now accept "0x..." syntax for backward compatibility.

	* Windows: The cfgmgr32 back-end which provides the list of devices
	  can be combined with another back-end which provides access
	  to configuration space.

	* ECAM (Enhanced Configuration Access Mechanism), which is defined
	  by the PCIe standard, is now supported. It requires root privileges,
	  access to physical memory, and also manual configuration on some
	  systems.

	* lspci: Tree view now works on multi-domain systems. It now respects
	  filters properly.

	* Last but not least, pci.ids were updated to the current snapshot
	  of the database. This includes overall cleanup of entries with
	  non-ASCII characters in their names -- such characters are allowed,
	  but only if they convey interesting information (e.g., umlauts
	  in German company names, but not the "registered trade mark" sign).

2022-11-20  Martin Mares <mj@ucw.cz>

	* Released as 3.9.0.

	* We decode Compute Express Link (CXL) capabilities.

	* The tree mode of lspci is now compatible with filtering options.

	* When setpci is used with a named register, it checks whether
	  the register is present in the particular header type.

	* Linux: The intel-conf[12] back-ends prefer to use ioperm() instead
	  of iopl() to gain access to I/O ports.

	* Windows: We have two new back-ends thanks to Pali Rohár.
	  One uses the NT SysDbg interface, the other uses kldbgdrv.sys
	  (which is a part of the Microsoft WinDbg tool).

	* Windows: We support building libpci as a DLL. Also, Windows
	  binaries now include meta-data with version.

	* Hurd: The Hurd back-end works again.

	* mmio-conf1(-ext): Added a new back-end implementing the intel-conf1
	  interface over MMIO. This is useful on some ARM machines, but it
	  requires manual configuration of the MMIO addresses.

	* As usually, updated pci.ids to the current snapshot of the database.

2022-04-18  Martin Mares <mj@ucw.cz>

	* Released as 3.8.0.

	* Filters can now match devices based on partially specified
	  class code and also on the programming interface.

	* Reporting of link speeds, power limits, and virtual function tags
	  has been updated to the current PCIe specification.

	* We decode the Data Object Exchange capability.

	* Bus mapping mode works in non-zero domains.

	* pci_fill_info() can fetch more fields: bridge bases, programming
	  interface, revision, subsystem vendor and device ID, OS driver,
	  and also parent bridge. Internally, the implementation was rewritten,
	  significantly reducing the number of corner cases to be handled.

	* The Windows port was revived and greatly improved by Pali Rohár.
	  It requires less magic to compile. More importantly, it runs on both
	  old and recent Windows systems (see README.Windows for details).

	* Added a new Windows back-end using the cfgmgr32 interface.
	  It does not provide direct access to the configuration space,

pcilmr.c
/*
 *	The PCI Utilities -- Margining utility main function
 *
 *	Copyright (c) 2023-2024 KNS Group LLC (YADRO)
 *
 *	Can be freely distributed and used under the terms of the GNU GPL v2+.
 *
 *	SPDX-License-Identifier: GPL-2.0-or-later
 */

#include <memory.h>
#include <stdio.h>
#include <stdlib.h>

#include "lmr/lmr.h"

const char program_name[] = "pcilmr";

static void
scan_links(struct pci_access *pacc, bool only_ready)
{
  if (only_ready)
    printf("Links ready for margining:\n");
  else
    printf("Links with Lane Margining at the Receiver capabilities:\n");
  bool flag = true;
  for (struct pci_dev *p = pacc->devices; p; p = p->next)
    {
      if (pci_find_cap(p, PCI_EXT_CAP_ID_LMR, PCI_CAP_EXTENDED) && margin_port_is_down(p))
        {
          struct pci_dev *down = NULL;
          struct pci_dev *up = NULL;
          margin_find_pair(pacc, p, &down, &up);

          if (down && margin_verify_link(down, up))
            {
              margin_log_bdfs(down, up);
              if (!only_ready && (margin_check_ready_bit(down) || margin_check_ready_bit(up)))
                printf(" - Ready");
              printf("\n");
              flag = false;
            }
        }
    }
  if (flag)
    printf("Links not found or you don't have enough privileges.\n");
  pci_cleanup(pacc);
  exit(0);
}

int
main(int argc, char **argv)
{
  struct pci_access *pacc;

  u8 links_n = 0;
  struct margin_link *links;
  bool *checks_status_ports;

  enum margin_mode mode;

  /* each link has several receivers -> several results */
  struct margin_results **results;
  u8 *results_n;

  pacc = pci_alloc();
  pci_init(pacc);
  pci_scan_bus(pacc);

  margin_print_domain = false;
  for (struct pci_dev *dev = pacc->devices; dev; dev = dev->next)
    {
      if (dev->domain != 0)
        {
          margin_print_domain = true;
          break;
        }
    }

  margin_global_logging = true;

  struct option long_options[]
    = { { .name = "margin", .has_arg = no_argument, .flag = NULL, .val = 0 },
        { .name = "scan", .has_arg = no_argument, .flag = NULL, .val = 1 },
        { .name = "full", .has_arg = no_argument, .flag = NULL, .val = 2 },
        { 0, 0, 0, 0 } };

  opterr = 0;
  int c;
  c = getopt_long(argc, argv, "+", long_options, NULL);

  switch (c)
    {
      case -1: /* no options (strings like component are possible) */
        /* FALLTHROUGH */
      case 0:
        mode = MARGIN;
        break;
      case 1:
        mode = SCAN;
        if (optind == argc)
          scan_links(pacc, false);
        else
          die("Invalid arguments\n\n%s", usage);
        break;
      case 2:
        mode = FULL;
        break;
      default: /* unknown option symbol */
        mode = MARGIN;
        optind--;
        break;
    }

  opterr = 1;

  links = margin_parse_util_args(pacc, argc, argv, mode, &links_n);
  struct margin_com_args *com_args = links[0].args.common;

  results = xmalloc(links_n * sizeof(*results));
  results_n = xmalloc(links_n * sizeof(*results_n));
  checks_status_ports = xmalloc(links_n * sizeof(*checks_status_ports));

  for (int i = 0; i < links_n; i++)
    {
      enum margin_test_status args_status;

      if ((args_status = margin_process_args(&links[i])) != MARGIN_TEST_OK)
        {
          checks_status_ports[i] = false;
          results[i] = xmalloc(sizeof(*results[i]));
          results[i]->test_status = args_status;
          continue;
        }

      checks_status_ports[i] = true;
      struct margin_params params;
      struct margin_link_args *link_args = &links[i].args;

      for (int j = 0; j < link_args->recvs_n; j++)
        {
          if (margin_read_params(
                pacc, link_args->recvs[j] == 6 ? links[i].up_port.dev : links[i].down_port.dev,
                link_args->recvs[j], &params))
            {
              u8 steps_t = link_args->steps_t ? link_args->steps_t : params.timing_steps;
              u8 steps_v = link_args->steps_v ? link_args->steps_v : params.volt_steps;
              u8 parallel_recv = link_args->parallel_lanes > params.max_lanes + 1 ?
                                   params.max_lanes + 1 :
                                   link_args->parallel_lanes;

              u8 step_multiplier
                = link_args->lanes_n / parallel_recv + ((link_args->lanes_n % parallel_recv) > 0);

              com_args->steps_utility += steps_t * step_multiplier;
              if (params.ind_left_right_tim)
                com_args->steps_utility += steps_t * step_multiplier;
              if (params.volt_support)
                {
                  com_args->steps_utility += steps_v * step_multiplier;
                  if (params.ind_up_down_volt)
                    com_args->steps_utility += steps_v * step_multiplier;
                }
            }
        }
    }

  for (int i = 0; i < links_n; i++)
    {
      if (checks_status_ports[i])
        results[i] = margin_test_link(&links[i], &results_n[i]);
      else
        {
          results_n[i] = 1;
          if (results[i]->test_status == MARGIN_TEST_ARGS_RECVS)
            {
              margin_log_link(&links[i]);
              printf("\nInvalid RecNums specified.\n");
            }
          else if (results[i]->test_status == MARGIN_TEST_ARGS_LANES)
            {
              margin_log_link(&links[i]);
              printf("\nInvalid lanes specified.\n");
            }
        }
      printf("\n----\n\n");
    }

  if (com_args->run_margin)
    {
      printf("Results:\n");
      printf(
        "Margining statuses:\nLIM -\tErrorCount exceeded Error Count Limit (found device limit)\n");
      printf("NAK -\tDevice didn't execute last command, \n\tso result may be less reliable\n");
      printf("THR -\tThe set (using the utility options) \n\tstep threshold has been reached\n\n");
      printf("Notations:\nst - steps\n\n");

      for (int i = 0; i < links_n; i++)
        {
          printf("Link ");
          margin_log_bdfs(links[i].down_port.dev, links[i].up_port.dev);
          printf(":\n\n");
          margin_results_print_brief(results[i], results_n[i], &links[i].args);
          if (com_args->save_csv)
            margin_results_save_csv(results[i], results_n[i], &links[i]);
          printf("\n");
        }
    }

  for (int i = 0; i < links_n; i++)
    margin_free_results(results[i], results_n[i]);
  free(results_n);
  free(results);
  free(com_args);
  free(links);
  free(checks_status_ports);

  pci_cleanup(pacc);
  return 0;
}
